src/uct/tcp/tcp_ep.c

```diff
@@ -1017,22 +1219,113 @@ ucs_status_t uct_tcp_ep_am_zcopy(uct_ep_h uct_ep, uint8_t am_id, const void *hea
-    if ((status == UCS_OK) || (status == UCS_ERR_NO_PROGRESS)) {
-        UCT_TL_EP_STAT_OP(&ep->super, AM, ZCOPY, hdr->length);
+    if (ucs_unlikely((status != UCS_OK) && (status != UCS_ERR_NO_PROGRESS))) {
+        goto out;
+    }
 
-        if (uct_tcp_ep_ctx_buf_need_progress(&ep->tx)) {
-            uct_tcp_ep_set_outstanding_zcopy(iface, ep, ctx, header,
-                                             header_length, comp);
-            return UCS_INPROGRESS;
-        }
+    UCT_TL_EP_STAT_OP(&ep->super, AM, ZCOPY, ctx->super.length);
+
+    if (uct_tcp_ep_ctx_buf_need_progress(&ep->tx)) {
+        uct_tcp_ep_set_outstanding_zcopy(iface, ep, ctx, header,
+                                         header_length, comp);
+        return UCS_INPROGRESS;
+    }
+
+    ucs_assert(status == UCS_OK);
+
+out:
+    uct_tcp_ep_ctx_reset(&ep->tx);
+    return status;
+}
+
+ucs_status_t uct_tcp_ep_put_zcopy(uct_ep_h uct_ep, const uct_iov_t *iov,
+                                  size_t iovcnt, uint64_t remote_addr,
+                                  uct_rkey_t rkey, uct_completion_t *comp)
+{
+    uct_tcp_ep_t *ep             = ucs_derived_of(uct_ep, uct_tcp_ep_t);
+    uct_tcp_iface_t *iface       = ucs_derived_of(uct_ep->iface, uct_tcp_iface_t);
+    uct_tcp_ep_zcopy_tx_t *ctx   = NULL;
+    uct_tcp_ep_put_hdr_t put_req = {0}; /* Suppress Cppcheck false-positive */
+    ucs_status_t status;
+
+    UCT_CHECK_LENGTH(sizeof(put_req) + uct_iov_total_length(iov, iovcnt), 0,
+                     UCT_TCP_EP_PUT_ZCOPY_MAX - sizeof(uct_tcp_am_hdr_t),
+                     "put_zcopy");
+
+    status = uct_tcp_ep_prepare_zcopy(iface, ep, UCT_TCP_EP_PUT_REQ_AM_ID,
+                                      &put_req, sizeof(put_req),
+                                      iov, iovcnt, "put_zcopy",
+                                      /* Set a payload length directly to the
+                                       * TX length, since PUT Zcopy doesn't
+                                       * set the payload length to TCP AM hdr */
+                                      &ep->tx.length, &ctx);
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    ctx->super.am_id  = UCT_TCP_EP_PUT_REQ_AM_ID;
+    ctx->super.length = sizeof(put_req);
+    put_req.addr      = remote_addr;
+    put_req.length    = ep->tx.length;
+
+    status = uct_tcp_ep_am_sendv(iface, ep, 0, &ctx->super, UCT_TCP_EP_PUT_ZCOPY_MAX,
+                                 &put_req, ctx->iov, ctx->iov_cnt);
+    if (ucs_unlikely((status != UCS_OK) && (status != UCS_ERR_NO_PROGRESS))) {
+        goto out;
+    }
+
+    UCT_TL_EP_STAT_OP(&ep->super, PUT, ZCOPY, put_req.length);
+
+    /* Increment ep::tx::unacked_put_sn in order to ensure returning
+     * UCS_INPROGRESS from flush function and do progressing.
+     * ep::tx::acked_put_sn has to be incremented upon PUT ACK
+     * message receiving */
+    uct_tcp_ep_commit_put_req(ep);
```

Comment:
imo, no need for this func - it is called from one place and just changes 2 counters (thru other inline funcs).
doing it right here would look simpler